testsupport/wait/host.go

```diff
@@ -162,138 +168,158 @@ func (a *HostAwaitility) UpdateUserSignupSpec(userSignupName string, modifyUserS
+			return test.ContainsCondition(actual.Status.Conditions, expected)
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected conditions to contain: %s.\n\tactual: %s", spew.Sdump(expected), spew.Sdump(actual.Status.Conditions))
+		},
 	}
 }
 
 // UntilMasterUserRecordHasConditions checks if MasterUserRecord status has the given set of conditions
-func UntilMasterUserRecordHasConditions(conditions ...toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
-	return func(a *HostAwaitility, mur *toolchainv1alpha1.MasterUserRecord) bool {
-		if test.ConditionsMatch(mur.Status.Conditions, conditions...) {
-			a.T.Logf("status conditions match in MasterUserRecord '%s`", mur.Name)
-			return true
-		}
-		a.T.Logf("waiting for status condition of MasterUserRecord '%s'. Actual: '%+v'; Expected: '%+v'", mur.Name, mur.Status.Conditions, conditions)
-		return false
+func UntilMasterUserRecordHasConditions(expected ...toolchainv1alpha1.Condition) MasterUserRecordWaitCriterion {
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			return test.ConditionsMatch(actual.Status.Conditions, expected...)
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected conditions to match:\n%s", Diff(actual.Status.Conditions, expected))
+		},
 	}
 }
 
 // UntilMasterUserRecordHasNotSyncIndex checks if MasterUserRecord has a
 // sync index *different* from the given value for the given target cluster
-func UntilMasterUserRecordHasNotSyncIndex(syncIndex string) MasterUserRecordWaitCriterion {
-	return func(a *HostAwaitility, mur *toolchainv1alpha1.MasterUserRecord) bool {
-		// lookup user account with target cluster
-		ua := mur.Spec.UserAccounts[0]
-		a.T.Logf("expecting sync indexes '%s' != '%s'", ua.SyncIndex, syncIndex)
-		return ua.SyncIndex != syncIndex
+func UntilMasterUserRecordHasNotSyncIndex(expected string) MasterUserRecordWaitCriterion {
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			// lookup user account with target cluster
+			ua := actual.Spec.UserAccounts[0]
+			return ua.SyncIndex != expected
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected sync index to match.\n\twanted: '%s'\n\tactual: '%s'", expected, actual.Spec.UserAccounts[0].SyncIndex)
+		},
 	}
 }
 
 func WithMurName(name string) MasterUserRecordWaitCriterion {
-	return func(a *HostAwaitility, mur *toolchainv1alpha1.MasterUserRecord) bool {
-		return mur.Name == name
+	return MasterUserRecordWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.MasterUserRecord) bool {
+			return actual.Name == name
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return "" // unused
```

Comment:
Why? We should print that there was no MUR with the given name found.

Reply:
the `Diff` method isn't called, but the caller (test) will report an error. I think that covers our needs.

Reply:
I thought that printing the diff or the error log is a responsibility of the criteria that wasn't met, not the responsibility of the caller. This is at least what I see in the other cases.

Reply:
right, my answer wasn't accurate. I'll add a message and display it

Reply:
done in 1641853